File: frontend/admin-panel/src/types/product.ts
// 商品相关TypeScript类型定义

export type ProductStatus = 'draft' | 'active' | 'inactive' | 'deleted';

export type CategoryStatus = 1 | 0; // 1: 启用, 0: 禁用

export type ChangeOperation = 'create' | 'update' | 'delete' | 'status_change';

export interface Money {
  amount: number; // 以分为单位
  currency: string;
}

export interface InventoryInfo {
  stock_quantity: number;
  reserved_quantity: number;
  track_inventory: boolean;
  low_stock_threshold?: number;
  reorder_point?: number;
  reorder_quantity?: number;
  cost_per_unit?: Money;
}

export interface ProductImage {
  id: string;
  url: string;
  alt_text?: string;
  sort_order: number;
  is_primary: boolean;
}

export interface Product {
  id: number;
  tenant_id: number;
  merchant_id: number;
  name: string;
  description?: string;
  category_id?: number;
  category_path?: string;
  tags: string[];
  price: Money;
  rights_cost: number;
  inventory: InventoryInfo;
  status: ProductStatus;
  images: ProductImage[];
  version: number;
  created_at: string;
  updated_at: string;
}

export interface ProductCategory {
  id: number;
  tenant_id: number;
  name: string;
  parent_id?: number;
  level: number;
  path: string;
  sort_order: number;
  status: CategoryStatus;
  created_at: string;
  updated_at: string;
}

export interface ProductHistory {
  id: number;
  tenant_id: number;
  product_id: number;
  version: number;
  field_name: string;
  old_value?: string;
  new_value?: string;
  operation: ChangeOperation;
  changed_by: number;
  changed_at: string;
}

// API请求类型
export interface CreateProductRequest {
  name: string;
  description?: string;
  category_id?: number;
  tags?: string[];
  price: Money;
  rights_cost: number;
  inventory: InventoryInfo;
}

export interface UpdateProductRequest {
  name?: string;
  description?: string;
  category_id?: number;
  tags?: string[];
  price?: Money;
  rights_cost?: number;
  inventory?: InventoryInfo;
}

export interface UpdateProductStatusRequest {
  status: ProductStatus;
}

export interface ProductListRequest {
  page: number;
  page_size: number;
  category_id?: number;
  status?: ProductStatus;
  keyword?: string;
  sort_by?: 'created_at' | 'updated_at' | 'name' | 'price';
  sort_order?: 'asc' | 'desc';
}

export interface ProductBatchOperationRequest {
  product_ids: number[];
  operation: 'activate' | 'deactivate' | 'delete';
}

export interface CreateCategoryRequest {
  name: string;
  parent_id?: number;
  sort_order?: number;
}

export interface UpdateCategoryRequest {
  name?: string;
  parent_id?: number;
  sort_order?: number;
  status?: CategoryStatus;
}

export interface UploadImageRequest {
  alt_text?: string;
  sort_order?: number;
  is_primary?: boolean;
}

// API响应类型
export interface ProductResponse {
  product: Product;
  category?: ProductCategory;
}

export interface ProductListResponse {
  products: ProductResponse[];
  total: number;
  page: number;
  page_size: number;
}

export interface CategoryTreeResponse extends ProductCategory {
  children?: CategoryTreeResponse[];
}

// Amis Schema相关类型
export interface AmisSchema {
  type: string;
  [key: string]: any;
}

export interface ProductListSchema extends AmisSchema {
  type: 'crud';
  api: string;
  columns: Array<{
    name: string;
    label: string;
    type?: string;
    [key: string]: any;
  }>;
}

export interface ProductFormSchema extends AmisSchema {
  type: 'form';
  api: string;
  controls: Array<{
    type: string;
    name: string;
    label: string;
    [key: string]: any;
  }>;
}

// 库存管理相关类型定义
export type InventoryChangeType = 'purchase' | 'sale' | 'adjustment' | 'transfer' | 'damage' | 'reservation' | 'release';
export type InventoryAlertType = 'low_stock' | 'out_of_stock' | 'overstock';
export type ReservationStatus = 'active' | 'confirmed' | 'released' | 'expired';

export interface InventoryRecord {
  id: number;
  tenant_id: number;
  product_id: number;
  change_type: InventoryChangeType;
  quantity_before: number;
  quantity_after: number;
  quantity_changed: number;
  reason: string;
  reference_id?: string;
  operated_by: number;
  created_at: string;
}

export interface InventoryReservation {
  id: number;
  tenant_id: number;
  product_id: number;
  reserved_quantity: number;
  reference_type: string;
  reference_id: string;
  status: ReservationStatus;
  expires_at?: string;
  created_at: string;
  updated_at: string;
}

export interface InventoryAlert {
  id: number;
  tenant_id: number;
  product_id: number;
  alert_type: InventoryAlertType;
  threshold_value: number;
  notification_channels: string[];
  is_active: boolean;
  last_triggered_at?: string;
  created_at: string;
  updated_at: string;
}

export interface InventoryStatistics {
  tenant_id: number;
  total_products: number;
  low_stock_products: number;
  out_of_stock_products: number;
  total_inventory_value: number;
  active_alerts: number;
  today_changes: number;
  last_updated: string;
}

// 库存操作请求类型
export interface InventoryAdjustRequest {
  product_id: number;
  adjustment_type: 'increase' | 'decrease' | 'set';
  quantity: number;
  reason: string;
  reference_id?: string;
}

export interface BatchInventoryAdjustRequest {
  adjustments: InventoryAdjustRequest[];
  reason: string;
}

export interface InventoryReserveRequest {
  product_id: number;
  quantity: number;
  reference_type: string;
  reference_id: string;
  expires_at?: string;
}

export interface InventoryAlertRequest {
  product_id: number;
  alert_type: InventoryAlertType;
  threshold_value: number;
  notification_channels: string[];
  is_active: boolean;
}

// 库存响应类型
export interface InventoryResponse {
  product_id: number;
  inventory_info: InventoryInfo;
  available_stock: number;
  reserved_stock: number;
  is_low_stock: boolean;
  is_out_of_stock: boolean;
}

export interface InventoryRecordResponse {
  records: InventoryRecord[];
  total: number;
  page: number;
  page_size: number;
}

export interface InventoryMonitoringData {
  statistics: InventoryStatistics;
  recent_changes: InventoryRecord[];
  active_alerts: InventoryAlert[];
  last_updated: string;
}

// 库存盘点相关类型定义
export type StocktakingStatus = 'pending' | 'in_progress' | 'completed' | 'cancelled';

export interface InventoryStocktaking {
  id: number;
  tenant_id: number;
  name: string;
  description: string;
  status: StocktakingStatus;
  product_ids: number[]; // 如果为空则盘点所有商品
  started_by: number;
  completed_by?: number;
  started_at: string;
  completed_at?: string;
  summary: string;
  notes: string;
  created_at: string;
  updated_at: string;
}

export interface InventoryStocktakingRecord {
  id: number;
  tenant_id: number;
  stocktaking_id: number;
  product_id: number;
  system_count: number;    // 系统库存数量
  actual_count: number;    // 实际盘点数量
  difference: number;      // 差异数量
  reason: string;          // 差异原因
  checked_by: number;      // 盘点人
  checked_at: string;      // 盘点时间
  created_at: string;
  updated_at: string;
}

// 盘点相关请求类型
export interface StocktakingCreateRequest {
  name: string;
  description?: string;
  product_ids?: number[]; // 如果为空则盘点所有商品
  start_time?: string;    // 盘点开始时间，为空则立即开始
}

export interface StocktakingRecordRequest {
  product_id: number;
  actual_count: number;
  system_count: number;
  reason?: string;
}

export interface StocktakingBatchUpdateRequest {
  records: StocktakingRecordRequest[];
}

export interface StocktakingCompleteRequest {
  summary?: string;
  notes?: string;
}

// 盘点响应类型
export interface StocktakingResponse {
  stocktaking: InventoryStocktaking;
  records?: InventoryStocktakingRecord[];
  statistics?: StocktakingStatistics;
}

export interface StocktakingStatistics {
  total_products: number;   // 总盘点商品数
  checked_products: number; // 已盘点商品数
  difference_count: number; // 有差异的商品数
  total_difference: number; // 总差异数量
}

export interface StocktakingListResponse {
  stocktakings: InventoryStocktaking[];
  total: number;
  page: number;
  page_size: number;
}

File: backend/shared/types/product.go
package types

import (
	"database/sql/driver"
	"encoding/json"
	"time"
)

// 扩展已有的ProductStatus枚举，添加DELETED状态
const (
	ProductStatusDeleted ProductStatus = "deleted" // 已删除
)

// CategoryStatus 分类状态枚举
type CategoryStatus int

const (
	CategoryStatusActive   CategoryStatus = 1 // 启用
	CategoryStatusInactive CategoryStatus = 0 // 禁用
)

// ChangeOperation 变更操作类型
type ChangeOperation string

const (
	ChangeOperationCreate      ChangeOperation = "create"
	ChangeOperationUpdate      ChangeOperation = "update"
	ChangeOperationDelete      ChangeOperation = "delete"
	ChangeOperationStatusChange ChangeOperation = "status_change"
)

// ProductImage 商品图片
type ProductImage struct {
	ID        string `json:"id"`
	URL       string `json:"url"`
	AltText   string `json:"alt_text,omitempty"`
	SortOrder int    `json:"sort_order"`
	IsPrimary bool   `json:"is_primary"`
}

// ProductImages 商品图片数组类型，实现数据库序列化
type ProductImages []ProductImage

// Value 实现 driver.Valuer 接口
func (p ProductImages) Value() (driver.Value, error) {
	if len(p) == 0 {
		return json.Marshal([]ProductImage{})
	}
	return json.Marshal(p)
}

// Scan 实现 sql.Scanner 接口
func (p *ProductImages) Scan(value interface{}) error {
	if value == nil {
		*p = ProductImages{}
		return nil
	}
	
	switch v := value.(type) {
	case []byte:
		return json.Unmarshal(v, p)
	case string:
		return json.Unmarshal([]byte(v), p)
	}
	return nil
}

// StringArray 字符串数组类型，用于商品标签
type StringArray []string

// Value 实现 driver.Valuer 接口
func (s StringArray) Value() (driver.Value, error) {
	if len(s) == 0 {
		return json.Marshal([]string{})
	}
	return json.Marshal(s)
}

// Scan 实现 sql.Scanner 接口
func (s *StringArray) Scan(value interface{}) error {
	if value == nil {
		*s = StringArray{}
		return nil
	}
	
	switch v := value.(type) {
	case []byte:
		return json.Unmarshal(v, s)
	case string:
		return json.Unmarshal([]byte(v), s)
	}
	return nil
}

// EnhancedProduct 增强的商品结构，扩展原有Product
type EnhancedProduct struct {
	ID           uint64         `json:"id" db:"id"`
	TenantID     uint64         `json:"tenant_id" db:"tenant_id"`
	MerchantID   uint64         `json:"merchant_id" db:"merchant_id"`
	Name         string         `json:"name" db:"name"`
	Description  string         `json:"description" db:"description"`
	
	// 扩展字段
	CategoryID   *uint64       `json:"category_id,omitempty" db:"category_id"`
	CategoryPath string        `json:"category_path,omitempty" db:"category_path"`
	Tags         StringArray   `json:"tags" db:"tags"`
	
	// 价格信息 - 使用JSON格式存储
	Price        json.RawMessage `json:"price" db:"price"` // 存储Money JSON
	RightsCost   int64          `json:"rights_cost" db:"rights_cost"` // 以分为单位
	
	// 库存信息 - 继续使用原有格式
	InventoryInfo *InventoryInfo `json:"inventory_info" db:"inventory_info"`
	
	// 状态和图片
	Status       ProductStatus  `json:"status" db:"status"`
	Images       ProductImages  `json:"images" db:"images"`
	Version      int           `json:"version" db:"version"`
	
	CreatedAt    time.Time     `json:"created_at" db:"created_at"`
	UpdatedAt    time.Time     `json:"updated_at" db:"updated_at"`
}

// GetPriceMoney 获取价格Money结构
func (ep *EnhancedProduct) GetPriceMoney() (*Money, error) {
	var money Money
	err := json.Unmarshal(ep.Price, &money)
	if err != nil {
		return nil, err
	}
	return &money, nil
}

// SetPriceMoney 设置价格Money结构
func (ep *EnhancedProduct) SetPriceMoney(money *Money) error {
	priceJSON, err := json.Marshal(money)
	if err != nil {
		return err
	}
	ep.Price = priceJSON
	return nil
}

// ToLegacyProduct 转换为原有Product结构
func (ep *EnhancedProduct) ToLegacyProduct() *Product {
	money, _ := ep.GetPriceMoney()
	var priceAmount float64
	var priceCurrency string
	
	if money != nil {
		priceAmount = float64(money.Amount) / 100 // 转换为元
		priceCurrency = money.Currency
	}
	
	return &Product{
		ID:            ep.ID,
		TenantID:      ep.TenantID,
		MerchantID:    ep.MerchantID,
		Name:          ep.Name,
		Description:   ep.Description,
		PriceAmount:   priceAmount,
		PriceCurrency: priceCurrency,
		RightsCost:    float64(ep.RightsCost) / 100, // 转换为元
		InventoryInfo: ep.InventoryInfo,
		Status:        ep.Status,
		CreatedAt:     ep.CreatedAt,
		UpdatedAt:     ep.UpdatedAt,
	}
}

// ProductCategory 商品分类实体
type ProductCategory struct {
	ID        uint64         `json:"id" gorm:"primaryKey;autoIncrement"`
	TenantID  uint64         `json:"tenant_id" gorm:"not null;index:idx_tenant_parent"`
	Name      string         `json:"name" gorm:"not null;size:100"`
	ParentID  *uint64        `json:"parent_id,omitempty" gorm:"index:idx_tenant_parent"`
	Level     int            `json:"level" gorm:"not null;default:1;index:idx_level"`
	Path      string         `json:"path" gorm:"not null;size:500;index:idx_path"`
	SortOrder int            `json:"sort_order" gorm:"default:0"`
	Status    CategoryStatus `json:"status" gorm:"not null;default:1"`
	CreatedAt time.Time      `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt time.Time      `json:"updated_at" gorm:"autoUpdateTime"`
}

// TableName 指定表名
func (ProductCategory) TableName() string {
	return "product_categories"
}

// ProductHistory 商品变更历史实体
type ProductHistory struct {
	ID        uint64          `json:"id" gorm:"primaryKey;autoIncrement"`
	TenantID  uint64          `json:"tenant_id" gorm:"not null;index:idx_tenant_product"`
	ProductID uint64          `json:"product_id" gorm:"not null;index:idx_tenant_product"`
	Version   int             `json:"version" gorm:"not null;index:idx_version"`
	FieldName string          `json:"field_name" gorm:"not null;size:100"`
	OldValue  string          `json:"old_value,omitempty" gorm:"type:text"`
	NewValue  string          `json:"new_value,omitempty" gorm:"type:text"`
	Operation ChangeOperation `json:"operation" gorm:"not null;size:20"`
	ChangedBy uint64          `json:"changed_by" gorm:"not null"`
	ChangedAt time.Time       `json:"changed_at" gorm:"autoCreateTime;index:idx_changed_at"`
}

// TableName 指定表名
func (ProductHistory) TableName() string {
	return "product_histories"
}

// CreateProductRequest 创建商品请求
type CreateProductRequest struct {
	Name         string        `json:"name" validate:"required,max=255"`
	Description  string        `json:"description,omitempty"`
	CategoryID   *uint64       `json:"category_id,omitempty"`
	Tags         []string      `json:"tags,omitempty"`
	Price        Money         `json:"price" validate:"required"`
	RightsCost   int64         `json:"rights_cost" validate:"min=0"`
	Inventory    InventoryInfo `json:"inventory" validate:"required"`
}

// UpdateProductRequest 更新商品请求
type UpdateProductRequest struct {
	Name        string        `json:"name,omitempty" validate:"max=255"`
	Description string        `json:"description,omitempty"`
	CategoryID  *uint64       `json:"category_id,omitempty"`
	Tags        []string      `json:"tags,omitempty"`
	Price       *Money        `json:"price,omitempty"`
	RightsCost  *int64        `json:"rights_cost,omitempty" validate:"min=0"`
	Inventory   *InventoryInfo `json:"inventory,omitempty"`
}

// UpdateProductStatusRequest 更新商品状态请求
type UpdateProductStatusRequest struct {
	Status ProductStatus `json:"status" validate:"required,oneof=draft active inactive deleted"`
}

// ProductListRequest 商品列表查询请求
type ProductListRequest struct {
	Page       int           `json:"page" validate:"min=1"`
	PageSize   int           `json:"page_size" validate:"min=1,max=100"`
	CategoryID *uint64       `json:"category_id,omitempty"`
	Status     ProductStatus `json:"status,omitempty"`
	Keyword    string        `json:"keyword,omitempty"`
	SortBy     string        `json:"sort_by,omitempty" validate:"oneof=created_at updated_at name price"`
	SortOrder  string        `json:"sort_order,omitempty" validate:"oneof=asc desc"`
}

// ProductBatchOperationRequest 批量操作请求
type ProductBatchOperationRequest struct {
	ProductIDs []uint64      `json:"product_ids" validate:"required,min=1"`
	Operation  string        `json:"operation" validate:"required,oneof=activate deactivate delete"`
}

// CreateCategoryRequest 创建分类请求
type CreateCategoryRequest struct {
	Name      string  `json:"name" validate:"required,max=100"`
	ParentID  *uint64 `json:"parent_id,omitempty"`
	SortOrder int     `json:"sort_order,omitempty"`
}

// UpdateCategoryRequest 更新分类请求
type UpdateCategoryRequest struct {
	Name      string  `json:"name,omitempty" validate:"max=100"`
	ParentID  *uint64 `json:"parent_id,omitempty"`
	SortOrder *int    `json:"sort_order,omitempty"`
	Status    *CategoryStatus `json:"status,omitempty"`
}

// UploadImageRequest 上传图片请求
type UploadImageRequest struct {
	AltText   string `json:"alt_text,omitempty"`
	SortOrder int    `json:"sort_order,omitempty"`
	IsPrimary bool   `json:"is_primary,omitempty"`
}

// ProductResponse 商品响应
type ProductResponse struct {
	Product
	Category *ProductCategory `json:"category,omitempty"`
}

// ProductListResponse 商品列表响应
type ProductListResponse struct {
	Products []ProductResponse `json:"products"`
	Total    int64             `json:"total"`
	Page     int               `json:"page"`
	PageSize int               `json:"page_size"`
}

// CategoryTreeResponse 分类树响应
type CategoryTreeResponse struct {
	ProductCategory
	Children []CategoryTreeResponse `json:"children,omitempty"`
}